Add draft README for Kubernetes Workload Registrar tutorial (#59)
spiffe/spire-tutorials: k8s/k8s-workload-registrar/README.md

 * The steps in this document should work with Kubernetes version 1.20.2.

We will deploy an scenario that consists of a StatefulSet containing a SPIRE Server and the Kubernetes Workload Registrar, a SPIRE Agent, and a workload, and configure the different modes to illustrate the automatic registration entries creation.

# Common configuration

Below, we describe parts of the configurations that are common to the three modes.

The SPIRE Server and the Kubernetes Workload registrar will communicate each other using a socket, that will be mounted at the `/tmp/spire-server/private` directory, as we can see from the `volumeMounts` section of both containers. The only difference between these sections is that, for the registrar, the socket will have the `readOnly` option set to `true`, while for the SPIRE Server container it will have its value set to `false`. The registrar container's section that illustrate what we described earlier is shown:

```
volumeMounts:
- name: spire-server-socket
  mountPath: /tmp/spire-server/private
  readOnly: true
```

# Webhook mode (default)

> 

Reviewer: Yes, I left part of this change out so I can see why you are confused. I am proposing that we demote the descriptions of the modes under a new heading. So currently we have this outline (sorry, some of these names may be old):

```
Configure SPIRE to use the Kubernetes Workload Registrar
Prerequisites
Common configuration
Webhook mode (default)
  Pod deletion
  Teardown
Reconcile mode
  Pod deletion
  Non-labeled pods
  Teardown
CRD mode
  Pod deletion
  Non-annotated pods
  SPIFFE ID CRD creation
  SPIFFE ID CRD deletion
  Teardown
```

I'm proposing that we demote the three different modes under a new heading called "Configure workload registrar modes" like this:

```
Configure SPIRE to use the Kubernetes Workload Registrar
Prerequisites
Common configuration
Configure workload registrar modes
  Webhook mode (default)
    Pod deletion
    Teardown
  Reconcile mode
    Pod deletion
    Non-labeled pods
    Teardown
  CRD mode
    Pod deletion
    Non-annotated pods
    SPIFFE ID CRD creation
    SPIFFE ID CRD deletion
    Teardown
```

So then those new short sentences I added would be inserted under the start of each mode heading.

However, I don't feel too strongly about this, so we can leave the outline as is if you prefer.

> 

Author: Oh, I see know what you were meaning. Still, the proposed outline don't quite fit with the tutorial. I think that a good structure could be:

> 

Reviewer: 👍 Yes, I think it's better than my version 😃 . I have one request, though. Please include at least one sentence between each heading. For example, under the new heading "Configure Webhook mode" have a sentence like the one I put before: "This section describes the older, default webhook mode of the Kubernetes Workload Registrar." Then put the new "Run the registrar in Webhook mode" heading.

> 

Author: Sounds good, will add it.

```

To mount the two files into the node, we will use the special option `extraMounts` of kind, that allows us to pass files between the host and a kind node. 

```
nodes:
- role: control-plane
  extraMounts:
  - containerPath: /etc/kubernetes/pki/admctrl
    hostPath: WEBHOOKDIR/k8s/admctrl
```

The ```WEBHOOKDIR/k8s/admctrl``` variable points to the folder host path where the files are stored, and `containerPath` specify the directory in which the files will be mounted.

In kind's specific configuration file (`kind-config.yaml`), we use the `admission-control-config-file` option to tell the API server where to find the admission configuration. Note that this value match with the one set in `ContainerPath`.

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
kubeadmConfigPatches:
- |
  apiVersion: kubeadm.k8s.io/v1beta2
  kind: ClusterConfiguration
  metadata:
    name: config
  apiServer:
    extraArgs:
        "service-account-signing-key-file": "/etc/kubernetes/pki/sa.key"
        "service-account-issuer": "api"
        "service-account-api-audiences": "api,spire-server"
        "admission-control-config-file": "/etc/kubernetes/pki/admctrl/admission-control.yaml"
...
```

> 

Author: I like the proposed title (Run the registrar in webhook mode)

```

We omitted the entry ids, as it may change with every run. Let's see how the other fields are built:

The cluster name *example-cluster* is used as Parent ID for the entries that correspond to the agent and the workload, but there is no reference to the node that this pods belong to, this is, the registration entries are mapped to a single node entry inside the cluster. This represents a drawback for this mode, as all the nodes in the cluster have permission to get identities for all the workloads that belong to the Kubernetes cluster, which increases the blast radius in case of a node being compromised, among other disadvantages.

Taking a look on the assigned SPIFFE IDs for the agent and the workload, we can see that they have the following form:
*spiffe://\<TRUSTDOMAIN\>/ns/\<NAMESPACE\>/sa/\<SERVICEACCOUNT\>*.
From this, we can conclude that we are using the registrar configured with the Service Account Based workload registration (which is the default behaviour). For instance, as the workload uses the *default* service account, into the *spire* namespace, its SPIFFE ID is: *spiffe://example.org/ns/spire/sa/default* 

> Another thing that is worth to examine is the registrar log, in which we will found out if the entries were created by this container. Run the following command to get the logs of the registrar, and to look for the *Created pod entry* keyword. 

Author: Using the command the line below, we are looking for "evidence" that the registrar created the entries, and not the entries itself. What about something like:

> Another thing that is worth to examine is the registrar log, in which we will found out if the entries were created by this container. Run the following command to get the logs of the registrar, and to look for the *Created pod entry* keyword. 

Author: By issuing the command we are doing two things: getting the logs of the registrar, and looking over this text using the grep tool. Having this in mind, your suggestion still apply?

> Another thing that is worth to examine is the registrar log, in which we will found out if the entries were created by this container. Run the following command to get the logs of the registrar, and to look for the *Created pod entry* keyword. 

Reviewer: It reads a little funny to phrase it that way though it isn't wrong. I think the majority of readers will know what grep does so it's probably not necessary to emphasize the two steps that occur (get log, then filter log). So maybe we can rewrite it like this? If you really disagree, it is ok to use your version.

```
Run the following command to display lines in the log that match *Created pod entry*.
```

> Another thing that is worth to examine is the registrar log, in which we will found out if the entries were created by this container. Run the following command to get the logs of the registrar, and to look for the *Created pod entry* keyword. 

Author: I like that way, but i'd slightly change it to:

```console
Found 2 entries
Entry ID         : ...
SPIFFE ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node
Parent ID        : spiffe://example.org/spire/server
Revision         : 0
TTL              : default
Selector         : k8s_psat:cluster:example-cluster

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/ns/spire/sa/spire-agent
Parent ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node
Revision         : 0
TTL              : default
Selector         : k8s:ns:spire
Selector         : k8s:pod-name:spire-agent-wtx7b
```

We will check the registrar logs to find out if the registrar deleted the entry, looking for the "Deleted pod entry" keyword, with the command shown below:

```console 
$ kubectl logs statefulset/spire-server -n spire -c k8s-workload-registrar | grep "Deleting pod entries"
```

The registrar successfuly deleted the corresponding entry for the *example-workload* pod.

## Teardown

To delete the resources used for this mode, we will delete the cluster by executing:

```console
kind delete cluster --name example-cluster
```

# Reconcile mode

> This mode, as opposed to Webhook mode, does not use a validating webhook but two reconciling controllers instead: one for the nodes and one for the pods. For this reason, we will not deploy all the configuration needed to perform the Kubernetes API server authentication, as the secret and `KubeConfig` entry, for instance, situation that makes the configuration simpler. 

Author: The secret and kubeconfig entry are examples of files that are used to configure the API server authentication (and not the authentication itself). Maybe we can add something like:

```

Your output should similar to the following, and shows the entries for the node, the agent and the workload:

```console
Found 3 entries
Entry ID         : ...
SPIFFE ID        : spiffe://example.org/spire-k8s-registrar/example-cluster/node/example-cluster-control-plane
Parent ID        : spiffe://example.org/spire/server
Revision         : 0
TTL              : default
Selector         : k8s_psat:agent_node_name:example-cluster-control-plane
Selector         : k8s_psat:cluster:example-cluster

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/agent
Parent ID        : spiffe://example.org/spire-k8s-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 0
TTL              : default
Selector         : k8s:ns:spire
Selector         : k8s:pod-name:spire-agent-c5c5f

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/example-workload
Parent ID        : spiffe://example.org/spire-k8s-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 0
TTL              : default
Selector         : k8s:ns:spire
Selector         : k8s:pod-name:example-workload-b98cc787d-kzxz6
```

> If we compare this entries to those created using Webhook mode, the difference is that the Parent ID of the SVID contains a reference to the node name where the pod is scheduled on, in this case, `example-cluster-control-plane`. We mentioned that this doesn't happen using the Webhook node, and this was one of its principal drawbacks. Also, for the node registration entry (the one that has the SPIRE Server SPIFFE ID as the Parent ID), the node name is used in the selectors, along with the cluster name. For the remaining two entries, pod name and namespace are used in the selectors instead.

Author: I think that a little rewriting is needed. What do you think about the following paragraph:

```suggestion
If we compare these entries to those created using webhook mode, the difference is that the Parent ID of the agent and workload SVIDs (second and third, respectively) contains a reference to the node where the pods are scheduled on, in this case, using its name `minikube`. We mentioned that this doesn't happen using the webhook mode, and this was one of the principal drawbacks of that mode. Also, pod name and namespace are used in the selectors. For the node registration entry (the one that has the SPIRE Server SPIFFE ID as the Parent ID), the node name is used in the selectors, along with the cluster name.
```

> If we compare this entries to those created using Webhook mode, the difference is that the Parent ID of the SVID contains a reference to the node name where the pod is scheduled on, in this case, `example-cluster-control-plane`. We mentioned that this doesn't happen using the Webhook node, and this was one of its principal drawbacks. Also, for the node registration entry (the one that has the SPIRE Server SPIFFE ID as the Parent ID), the node name is used in the selectors, along with the cluster name. For the remaining two entries, pod name and namespace are used in the selectors instead.

Author: You're right on that, will change _SVID_ to _registration entry_ here and elsewhere.

Parent ID        : spiffe://example.org/spire/server
Revision         : 0
TTL              : default
Selector         : k8s_psat:agent_node_name:example-cluster-control-plane
Selector         : k8s_psat:cluster:example-cluster
```

If we look for the *Deleted entry* keyword on the registrar logs, we will find out that the registrar deleted the entry. Issue the following command:

```console
$ kubectl logs statefulset/spire-server -n spire -c k8s-workload-registrar | grep "Deleted entry"
```

The pod controller successfuly deleted the entry.

## Non-labeled pods

As we are using Label Based workload registration, only pods that have the *spire-workload* label will have its registration entry automatically created. Let's deploy a pod that has no label with the command below, by executing the comand below, from the `mode-reconcile` directory:

```console
$ kubectl apply -f k8s/not-labeled-workload.yaml
```

Let's see the existing registration entries with the command:

```console
$ kubectl exec -it statefulset/spire-server -n spire -c spire-server -- bin/spire-server entry show -registrationUDSPath /tmp/spire-server/private/api.sock
```

> The output should remain constant compared to the one that we obtained in the *Pod deletion* section. This implies that the only two registration entries on the SPIRE Server corresponds to the labeled deployed resources. This is the expected behaviour, as only labeled pods will be considered by the workload registrar while using the Label Workload registration mode.

Author: Allow me to rework this paragraph:

```suggestion
The output should be the same as the output that we obtained in the *Pod deletion* section. This implies that the only two registration entries on the SPIRE Server correspond to the labeled deployed resources. This is the expected behaviour, as only labeled pods will be considered by the workload registrar while using the Label Workload registration mode.
```

The concept that I'm trying to write is: "The registrar only creates entries for pods that are using the correct label"

> The output should remain constant compared to the one that we obtained in the *Pod deletion* section. This implies that the only two registration entries on the SPIRE Server corresponds to the labeled deployed resources. This is the expected behaviour, as only labeled pods will be considered by the workload registrar while using the Label Workload registration mode.

Author: Let's shorten this a little bit:

```suggestion
The output should be the same as the output that we obtained in the *Pod deletion* section. This implies that the registrar only creates entries for pods that are using the matching label.
```

I think it's better this way.

> The output should remain constant compared to the one that we obtained in the *Pod deletion* section. This implies that the only two registration entries on the SPIRE Server corresponds to the labeled deployed resources. This is the expected behaviour, as only labeled pods will be considered by the workload registrar while using the Label Workload registration mode.

Reviewer: 👍 Sounds great

## Teardown

To delete the resources used for this mode, we will delete the cluster by executing:

```console
kind delete cluster --name example-cluster
```

# CRD mode

This mode takes advantage of the `CustomResourceDefinition` feature from Kubernetes, which allows SPIRE to integrate with this tool and its control plane. A SPIFFE ID is defined as a custom resource, with an structure that matches the form of a registration entry. Below is a reduced example of the definition of a SPIFFE ID CRD.

```
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: my-test-spiffeid
  namespace: default
spec:
  parentId: spiffe://example.org/spire/server
  selector:
    namespace: default
    podName: my-test-pod
  spiffeId: spiffe://example.org/test
```

> The main goal of the custom resource is to track the intent of what and how the registration entries should look on the SPIRE Server, and to track any modification of these registration entries, reconciling its existence. This means that every SPIFFE ID CRD will have a matching registration entry, whose existence will be closely linked. Every modification done to the registration entry will have an impact on its corresponding SPIFFE ID CRD, and viceversa.

Author: Yep, it's a tricky word. What about modifying the pararaph to:

```suggestion
The main goal of the custom resource is to track the intent of what and how the registration entries should look on the SPIRE Server, by keeping this resources in sync with any modification made to the registration entries. This means that every SPIFFE ID CRD will have a matching registration entry, whose existence will be closely linked. Every modification done to a registration entry will have an impact on its corresponding SPIFFE ID CRD, and vice versa.
```

> The main goal of the custom resource is to track the intent of what and how the registration entries should look on the SPIRE Server, and to track any modification of these registration entries, reconciling its existence. This means that every SPIFFE ID CRD will have a matching registration entry, whose existence will be closely linked. Every modification done to the registration entry will have an impact on its corresponding SPIFFE ID CRD, and viceversa.

Reviewer: Thanks, your version works. But in the first sentence, please change the middle to: "...keeping these resources..." Also, please remove all of the commas.

Parent ID        : spiffe://example.org/spire/server
Revision         : 1
TTL              : default
Selector         : k8s_psat:agent_node_uid:08990bfd-3551-4761-8a1b-2e652984ffdd
Selector         : k8s_psat:cluster:example-cluster

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/testing/agent
Parent ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 1
TTL              : default
Selector         : k8s:node-name:example-cluster-control-plane
Selector         : k8s:ns:spire
Selector         : k8s:pod-uid:538886bb-48e1-4795-b386-10e97f50e34f
DNS name         : spire-agent-jzc8w

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/testing/example-workload
Parent ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 1
TTL              : default
Selector         : k8s:node-name:example-cluster-control-plane
Selector         : k8s:ns:spire
Selector         : k8s:pod-uid:78ed3fc5-4cff-476a-90f5-37d3abd47823
DNS name         : example-workload-6877cd47d5-l4hv5
```

> 3 entries were created corresponding to the node, agent, and workload. For the node entry (the one that has the SPIRE Server SPIFFE ID as Parent ID), we see a difference in the selectors, comparing it with the selectors in the node entry created using Reconcile mode for the same pod: we find out that instead of using the node name, CRD mode stores the UID of the node where the agent is running on. As the node name is used in the SPIFFE ID assigned to the node, we can take this as a mapping from node UID to node name.

Reviewer: ```suggestion
Three entries were created corresponding to the node, agent, and workload. For the node entry (the one that has the SPIRE Server SPIFFE ID as Parent ID), we see a difference in the selectors compared to reconcile mode: instead of using the node name, CRD mode stores the UID of the node where the agent is running on. As the node name is used in the SPIFFE ID assigned to the node, we can use this as a mapping from node UID to node name.
```

Author: Here I tried to use "take" as a synonym of "interpret", because we don't actually _use_ this mapping.

Reviewer: Ok, I see how my version is wrong but I don't quite understand the point you're making. We can use your updated version.

Author: I see this "mapping" thing occurring when the server issues the SVID to the agent. The server uses the selectors to know the identity that it should issue, so in this case, verifying the node UID that is present in the selectors, the server issues an SVID with an SPIFFE ID that has the node name in it. This is where I'm understanding that a mapping happens (node UID `08990bfd-....` -> node name `minikube`).

> Something similar happens with the pod entries, but this time the pod UID where the workload is running is stored in the selectors, instead of the node UID.

Reviewer: ```suggestion
Something similar happens with the pod entries, but this time the pod UID where the workload is running is stored in the selectors instead of the node UID.
```

If we now focus our attention on the SPIFFE IDs assigned to the workloads, we see that it takes the form of *spiffe://\<TRUSTDOMAIN\>/\<ANNOTATIONVALUE\>*. By using Annotation Based workload registration, it is possible to freely set the SPIFFE ID path. In this case, for the workload, we set the annotation value to *example-workload*.

Obtain the registrar logs by issuing: 

```console
kubectl logs statefulset/spire-server -n spire -c k8s-workload-registrar | grep "Created entry"
```

> This will show that the registrar created the 3 entries into the SPIRE Server.

Reviewer: ```suggestion
This will show that the registrar created the three entries in the SPIRE Server.
```

> In addition to the SPIRE entries, the registrar in this mode is configure to create the corresponding custom resources. Let's check for this using a Kubernetes native command such as:

Reviewer: ```suggestion
In addition to the SPIRE entries, the registrar in this mode is configured to create the corresponding custom resources. Let's check for this using a Kubernetes native command such as:
```

```console
$ kubectl delete deployment/example-workload -n spire
```

And now, check the registration entries in the SPIRE Server by executing:

```console
$ kubectl exec -it statefulset/spire-server -n spire -c spire-server -- bin/spire-server entry show -registrationUDSPath /tmp/spire-server/private/api.sock
```

The output should look like:

```console
Found 2 entries
Entry ID         : ...
SPIFFE ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Parent ID        : spiffe://example.org/spire/server
Revision         : 1
TTL              : default
Selector         : k8s_psat:agent_node_uid:08990bfd-3551-4761-8a1b-2e652984ffdd
Selector         : k8s_psat:cluster:example-cluster

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/testing/agent
Parent ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 1
TTL              : default
Selector         : k8s:node-name:example-cluster-control-plane
Selector         : k8s:ns:spire
Selector         : k8s:pod-uid:538886bb-48e1-4795-b386-10e97f50e34f
DNS name         : spire-agent-jzc8w
```

> The only entries that should exist now are the ones that match the node and the SPIRE agent, because the workload one was deleted by the registrar, something that we can check if we examine the registrar logs, but this time looking for the keyword "Deleted entry".

Reviewer: ```suggestion
The only entries that should exist now are the ones that match the node and the SPIRE Agent, because the workload one was deleted by the registrar, something that we can check if we examine the registrar logs, but this time looking for the keyword "Deleted entry":
```

```console
$ kubectl logs statefulset/spire-server -n spire -c k8s-workload-registrar | grep -A 1 "Deleted entry"
```

> As the registrar handles the custom resources automatically, it also deleted the corresponding SPIFFE ID CRD, something that we can also check by querying the Kubernetes control plane (```kubectl get spiffeids -n spire```), command from which we will obtaing the following:

Reviewer: ```suggestion
As the registrar handles the custom resources automatically, it also deleted the corresponding SPIFFE ID CRD, something that we can also check by querying the Kubernetes control plane (`kubectl get spiffeids -n spire`), command which should display the following:
```

```console
NAME                            AGE
example-cluster-control-plane   41m
spire-agent-r86rz               40m
```

## Non-annotated pods

> Let's check if a pod that has no annotations its considered by the registrar. Deploy a new workload with this condition with the following command:

Reviewer: ```suggestion
Let's check if a pod that has no annotations is detected by the registrar. Deploy a new workload without annotations using the following command:
```

```console
$ kubectl apply -f k8s/not-annotated-workload.yaml
```

As in the previous section, let's see the registration entries that are present in the SPIRE Server:

```console
$ kubectl exec -it statefulset/spire-server -n spire -c spire-server -- bin/spire-server entry show -registrationUDSPath /tmp/spire-server/private/api.sock
```

The result of the command should be equal to the one shown in *Pod deletion* section, because no new entry has been created, as expected.

## SPIFFE ID CRD creation

> One of the benefits of using the CRD Mode is that we can manipulate the SPIFFE IDs as if they were resources inside Kubernetes environment, in other words using the *kubectl* command.

Reviewer: ```suggestion
One of the benefits of using the CRD mode is that we can manipulate the SPIFFE IDs as if they were resources inside Kubernetes environment, in other words using the `kubectl` command.
```

Let's create a new SPIFFE ID CRD by using:

```console
$ kubectl apply -f k8s/test_spiffeid.yaml
``` 

> We will check if it was created, consulting the custom resources with ```kubectl get spiffeids -n spire```, whose output will show the following: 

Reviewer: ```suggestion
We will check if it was created, consulting the custom resources with `kubectl get spiffeids -n spire`, the output of which should show the following: 
```

```

You'll get an output similar to this:

```console
Found 3 entries
Entry ID         : ...
SPIFFE ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Parent ID        : spiffe://example.org/spire/server
Revision         : 1
TTL              : default
Selector         : k8s_psat:agent_node_uid:08990bfd-3551-4761-8a1b-2e652984ffdd
Selector         : k8s_psat:cluster:example-cluster

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/test
Parent ID        : spiffe://example.org/spire/server
Revision         : 1
TTL              : default
Selector         : k8s:ns:spire
Selector         : k8s:pod-name:my-test-pod

Entry ID         : ...
SPIFFE ID        : spiffe://example.org/testing/agent
Parent ID        : spiffe://example.org/k8s-workload-registrar/example-cluster/node/example-cluster-control-plane
Revision         : 1
TTL              : default
Selector         : k8s:node-name:example-cluster-control-plane
Selector         : k8s:ns:spire
Selector         : k8s:pod-uid:538886bb-48e1-4795-b386-10e97f50e34f
DNS name         : spire-agent-jzc8w
```

> As we can see, a SPIFFE ID CRD creation triggers a registration entry creation on the SPIRE Server too. 

Reviewer: ```suggestion
As we can see, SPIFFE ID CRD creation triggers registration entry creation on the SPIRE Server, too. 
```

## SPIFFE ID CRD deletion

> The lifecycle of a SPIFFE ID CRD can be managed by Kubernetes, and has a direct impact on the corresponding registration entry stored in the SPIRE Server. We already see how a SPIFFE ID CRD creation activates a registration entry one. We will prove that the same applies for a CRD deletion.

Reviewer: ```suggestion
The lifecycle of a SPIFFE ID CRD can be managed by Kubernetes, and has a direct impact on the corresponding registration entry stored in the SPIRE Server. We already saw how SPIFFE ID CRD creation activates registration entry creation. We will prove that the same applies for a CRD deletion.
```
